### PR TITLE
Prevent access to repo during step execution

### PR DESCRIPTION
**Pull request by schustmi**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- Create `Environment` class to group some existing functionality
- Prevent access to repo during step execution
- Enable autocomplete of `dockerignore_file` in pipeline decorator
- Remove unused exception

### CODE REVIEW

File: src/zenml/pipelines/pipeline_decorator.py [hunk 1]
```diff
@@ -37,6 +37,7 @@ def pipeline(
     name: Optional[str] = None,
     enable_cache: bool = True,
     requirements_file: Optional[str] = None,
+    dockerignore_file: Optional[str] = None,
```
Comment (AlexejPenner) [inline src/zenml/pipelines/pipeline_decorator.py:40]: Wasn't this in already?

File: src/zenml/repository.py [hunk 2]
```diff
@@ -113,7 +114,16 @@ def __init__(self, root: Optional[Path] = None):
...
         Raises:
             RepositoryNotFoundError: If no ZenML repository directory is found.
+            RuntimeError: If trying to create a `Repository` instance while a
+                ZenML step is being executed.
         """
+        if Environment.currently_running_step():
+            raise RuntimeError(
+                "Unable to access repository during step execution. If you "
+                "require access to the artifact or metadata store, please use "
+                "a `StepContext` inside your step instead."
```
Comment (AlexejPenner) [inline src/zenml/repository.py:124]: Personally, I think we could add a tiny example here that tells the user how he can achieve the desired fucntionality. Alternatively we could also refer them to the corresponding docs: https://docs.zenml.io/guides/index/historic-runs#utilizing-stepcontext

**Review by AlexejPenner — APPROVED** [review]

File: src/zenml/pipelines/pipeline_decorator.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/pipelines/pipeline_decorator.py:40]: It was in the actual decorator, but missing in the `@overload` which (at least for me in PyCharm) prevented it from showing up as an autocomplete option

File: src/zenml/repository.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/repository.py:124]: That is a great idea, I think I'd probably go for the URL as that goes into way more detail. Now I just need an Exception name :D

Comment (AlexejPenner) [reply src/zenml/repository.py:124]: How about ForbiddenRepositoryAccessError ?

**Review by stefannica — COMMENTED** [review]
You're hitting the nail on the head by adding a runtime environment object for ZenML. I've been looking for something where we can store context information related to the current pipeline run and step for a while now and your idea couldn't come at a better time.

I left a few high-level suggestions, let me know what you think about them.

File: src/zenml/repository.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/repository.py]: Fixed

File: src/zenml/environment.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+import os
+import platform
+from typing import Any, Dict
+
+import distro
+
+
+class Environment:
```
[the remark this answers is not in the document]
Comment (htahir1) [reply src/zenml/environment.py:21]: Would this class be exposed in the public API? If yes, then there is a chance to meddle with it if its a Singleton that we might want to consider. If not, it makes sense to have it as a singleton

File: src/zenml/repository.py [hunk 4]
```diff
@@ -113,7 +115,17 @@ def __init__(self, root: Optional[Path] = None):
 
         Raises:
             RepositoryNotFoundError: If no ZenML repository directory is found.
+            ForbiddenRepositoryAccessError: If trying to create a `Repository`
+                instance while a ZenML step is being executed.
         """
+        if Environment.currently_running_step():
+            raise ForbiddenRepositoryAccessError(
+                "Unable to access repository during step execution. If you "
```
Comment (htahir1) [inline src/zenml/repository.py:123]: After we do this we should update the nba example @AlexejPenner

File: tests/unit/test_repository.py [hunk 5]
```diff
@@ -189,6 +190,17 @@ def test_repo_without_configuration_file_falls_back_to_empty_config(tmp_path):
         _ = repo.active_stack
 
 
+def test_creating_repository_instance_during_step_execution_fails(mocker):
+    """Tests that creating a Repository instance while a step is being executed
```
Comment (htahir1) [inline tests/unit/test_repository.py:194]: Great test

**Review by htahir1 — APPROVED** [review]
Alright, I think that looks good. This Environment class seems to be solving some problems!

**Review by bcdurak — APPROVED** [review]
As we discussed earlier, I really like this new concept of `Environment`s. Especially, the way that it has this hidden paradigm with the class variable indicating whether it has been accessed throughout a step execution! Amazing work!

**Comment by schustmi** [comment]
I made some updates according to Stefans suggestions:
- the `Environment` class is now a singleton
- We can temporarily set attributes on the Environment using a contextmanager 
```python
with Environment()._set_attributes(...):
    ...
```

Would be nice if some of you could review these specific changes and the corresponding tests

**Review by htahir1 — CHANGES_REQUESTED** [review]
Ok this is really good now @schustmi ! I would just have one request to add the `running_in_notebook` function to Environment

**Comment by schustmi** [comment]
@htahir1 Done :)

**Review by stefannica — CHANGES_REQUESTED** [review]
Added a couple of naming suggestions, otherwise looks good.

File: src/zenml/environment.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            return False
+
+    @staticmethod
+    def in_paperspace_gradient() -> bool:
+        """If the current python process is running in Paperspace Gradient."""
+        return "PAPERSPACE_NOTEBOOK_REPO_ID" in os.environ
+
+    @classmethod
+    @contextmanager
+    def _temporary_context(
```
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/environment.py]: Thanks, I integrated both your suggestions and like it much better now!

File: src/zenml/integrations/evidently/visualizers/evidently_visualizer.py [hunk 7]
```diff
@@ -51,7 +52,7 @@ def generate_facet(self, html_: str) -> None:
         Args:
             html_: HTML represented as a string.
         """
-        if self.running_in_notebook():
+        if Environment.in_jupyter_notebook():
```
Comment (htahir1) [inline src/zenml/integrations/evidently/visualizers/evidently_visualizer.py:55]: Probably makes sense to add `or Environment.in_colab` here? If that isn't needed then I wouldn't call this `in_jupyter_notebook` rather just `in_notebook`

**Review by htahir1 — APPROVED** [review]
LGTM. Thanks for adding the function. Just left one tiny suggestion again but feel free to disregard.

File: src/zenml/integrations/evidently/visualizers/evidently_visualizer.py (hunk 7 above)
Comment (schustmi) [reply src/zenml/integrations/evidently/visualizers/evidently_visualizer.py:55]: The visualizations worked in colab before right? Then it probably works in both cases, I will try later an either update the name or add the `or Environment.in_colab`

Comment (schustmi) [reply src/zenml/integrations/evidently/visualizers/evidently_visualizer.py]: Works for both, so I renamed it to `in_notebook`

**Review by stefannica — APPROVED** [review]

**Pull request merged**